File: src/main/java/com/mercateo/jsonschema/generictype/GenericArray.java
package com.mercateo.jsonschema.generictype;

import java.lang.reflect.GenericArrayType;
import java.lang.reflect.Type;
import java.util.Objects;

import com.googlecode.gentyref.GenericTypeReflector;

final class GenericArray<T> extends GenericTypeAbstract<T, GenericArrayType> {

    GenericArray(GenericArrayType arrayType, Class<T> rawType) {
        super(rawType, arrayType);
    }

    @Override
    public String getSimpleName() {
        return type.getTypeName();
    }

    @Override
    public GenericType<?> getContainedType() {
        return GenericType.of(GenericTypeReflector.getArrayComponentType(type), getRawType()
                .getComponentType());
    }

    @Override
    public GenericType<? super T> getSuperType() {
        return null;
    }

    @Override
    public boolean isIterable() {
        return true;
    }
}

File: src/main/java/com/mercateo/jsonschema/generictype/GenericArray.kt
package com.mercateo.jsonschema.generictype

import com.googlecode.gentyref.GenericTypeReflector
import java.lang.reflect.GenericArrayType

internal class GenericArray<T>(
    arrayType: GenericArrayType,
    rawType: Class<T>
) : GenericTypeAbstract<T, GenericArrayType>(rawType, arrayType) {

    override val simpleName: String
        get() = type.typeName

    override val name: String
        get() = type.typeName

    override val containedType: GenericType<Any>
        get() = GenericType.of(GenericTypeReflector.getArrayComponentType(type), rawType.componentType)

    override val superType: GenericType<Any>?
        get() = null

    override val isIterable: Boolean
        get() = true
}
